[fake_impls] fix max_seqlen return values in efficient_attention_forward

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120842
* #120682

To match the actual implementation, we should return the max_seqlen_q/k, not M, N, when in the sparse case

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L981-L996

Note that although the .cu file sets max_seqlen_k = 0 in the sparse case, it actually returns max_seqlen_k or N:

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L1224-L1231

Tests - added in the next PR (#102839, which also fixes other parts of the test_fake tests so that we can un-xfail them and actually run the tests)

davidberard98 committed:
[fake_impls] fix max_seqlen return values in efficient_attention_forward

To match the actual implementation, we should return the max_seqlen_q/k, not M, N, when in the sparse case

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L981-L996

Note that although the .cu file sets max_seqlen_k = 0 in the non-sparse case, it actually returns max_seqlen_k or N:

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L1224-L1231

[ghstack-poisoned]

davidberard98 committed:
Update on "[fake_impls] fix max_seqlen return values in efficient_attention_forward"


To match the actual implementation, we should return the max_seqlen_q/k, not M, N, when in the sparse case

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L981-L996

Note that although the .cu file sets max_seqlen_k = 0 in the non-sparse case, it actually returns max_seqlen_k or N:

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L1224-L1231

[ghstack-poisoned]

YuqingJ commented on torch/_subclasses/fake_impls.py:
```diff
@@ -803,6 +803,7 @@ def meta__efficient_attention_forward(fake_mode, func, *args, **kwargs):
     max_seqlen_q = kwargs["max_seqlen_q"]
+    max_seqlen_k = kwargs["max_seqlen_k"]
     compute_log_sumexp = kwargs["compute_log_sumexp"]
     # unused: bias, cu_seqlens_k, max_seqlen_k, dropout_p, custom_mask_type, scale, causal_diagonal, seqlen_k
```
nit: remove max_seqlen_k from the list?

YuqingJ commented on torch/_subclasses/fake_impls.py:
```diff
@@ -826,6 +827,7 @@ def convert_tensor(t, device):
     if cu_seqlens_q is not None:
         assert max_seqlen_q is not None
         actual_max_seqlen_q = max_seqlen_q
+    actual_max_seqlen_k = max_seqlen_k if max_seqlen_k is not None else N
```
nit: should we keep it consistent with line 826-829 to get max_seqlen?

drisspg replied:
since in the kernel the check is also gated by if cu_seqlens_q is not none, I think add the assert and stuff `actual_max_seqlen_k = max_seqlen_k` in there

davidberard98 replied:
the logic for handling max_seqlen_k is different from the logic for handling max_seqlen_q (comment in the PR description - it's this line: `max_seqlen_k_.has_value() ? max_seqlen_k_.value() : max_seqlen_k); `). I don't think there are any asserts on max_seqlen_k either.

Planning to leave it as unless I missed something - please LMK if so

YuqingJ commented:
Did you mean to say "Note that although the .cu file sets max_seqlen_k = 0 in the sparse case"?

drisspg commented on test/test_ops.py:
```diff
@@ -2162,6 +2162,8 @@ def map_to_fake(e):
                 ):
                     if not isinstance(fake_out, torch.Tensor):
                         self.assertTrue(not isinstance(real_out, torch.Tensor))
+                        if isinstance(fake_out, (bool, int, float)):
```
Maybe leave a comment here, my interpretation of this code is that, previously  "if the fake out is not a tensor, make sure the real out isn't a tensor, everything else we don't care about' but now we do care but only about certain things, why these things?

davidberard98 replied:
Switched it to check everything

davidberard98 committed:
Update on "[fake_impls] fix max_seqlen return values in efficient_attention_forward"


To match the actual implementation, we should return the max_seqlen_q/k, not M, N, when in the sparse case

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L981-L996

Note that although the .cu file sets max_seqlen_k = 0 in the sparse case, it actually returns max_seqlen_k or N:

https://github.com/pytorch/pytorch/blob/7e185277cd6fadada367fa6318beff9e4127570a/aten/src/ATen/native/transformers/cuda/attention.cu#L1224-L1231

Tests - added in the next PR (#102839, which also fixes other parts of the test_fake tests so that we can un-xfail them and actually run the tests)

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
